andela/boondocks-bn-backend
# #169257409 User Role Settings

#### What does this PR do?
This PR creates a role management system for the API

#### Description of Task to be completed?
- [x] setup a roles controller
- [x] setup roles routes
- [x] setup roles middleware
- [x] setup tests

#### How should this be manually tested?
- clone this branch to your workstation and navigate to the project folder using Terminal
- install dependencies with the command `npm install`
- update your database schema by running `npm run migrate:undo` and then `npm run migrate`
- run the seeder file to populate the database with a user with the 'super_administrator' role by running the command `npx sequelize-cli db:seed:all`. The user's details are as follows:
```
{ 
    email: 'super@administrator.com',
    password: '12345678',
}
``` 
Test 1
- Open Postman and make a login request using the 'super_administrator' credentials above to receive a token
- Append the token to the Authorization Header: Bearer Token
- Make a 'GET' request to the endpoint 'http://localhost:3000/api/v1/auth/user/'
- You will get a list of all registered users
- Signup a new user with the signup endpoint
- By default, this new user has a role 'requester'
- Attempt to retrieve the list of users again with this new (requester role) user, you will get an 'insufficient privileges' error

Test 2
- Using the 'super_administrator' user, change the role of the 'requester' user by making a 'PATCH' request on the endpoint 'http://localhost:3000/api/v1/auth/user/role' and make sure you append your token to the Authorization Header: Bearer Token
- Include the following details in the request body:
```
{
    email: theRequesterUserEmail,
    role: 'manager'
}
```
- You will receive a confirmation message that the role has been changed successfully
- Attempt the same action using an invalid token or a token from a 'non-super_administrator' user and you will get an error message

Test 3:
- Run integration tests using the command `npm test`

#### Any background context you want to provide?
We shall use this endpoint to restrict certain endpoints to users with specific roles

#### What are the relevant pivotal tracker stories?
[#169257409](https://www.pivotaltracker.com/story/show/169257409)

#### Screenshots (if appropriate)
<img width="1436" alt="allUsersSuccess" src="https://user-images.githubusercontent.com/17108099/69291056-aa06cc80-0c0a-11ea-8cd7-4fea96c0cfad.png">
<img width="1439" alt="changeRoleSuccess" src="https://user-images.githubusercontent.com/17108099/69291066-b5f28e80-0c0a-11ea-9baa-c50d90d328f9.png">

#### Questions:
N/A
